src/google/protobuf/map.h

```diff
@@ -548,7 +548,7 @@ class Map {
     !defined(GOOGLE_PROTOBUF_OS_NACL) && !defined(GOOGLE_PROTOBUF_OS_ANDROID)
     template<class NodeType, class... Args>
     void construct(NodeType* p, Args&&... args) {
-      new (static_cast<void*>(p)) NodeType(std::forward<Args>(args)...);
+      new (const_cast<void*>(static_cast<const void*>(p))) NodeType(std::forward<Args>(args)...);
```

Comment:
The issue is #1266 . The problem shortly is the following:
Clang drop the next error message:
```
In file included from google/protobuf/map_field.cc:31:
In file included from ./google/protobuf/map_field.h:40:
In file included from ./google/protobuf/map_field_lite.h:34:
./google/protobuf/map.h:551:12: error: static_cast from 'const google::protobuf::MapKey *' to 'void *' is not allowed
      new (static_cast<void*>(p)) NodeType(std::forward<Args>(args)...);
           ^~~~~~~~~~~~~~~~~~~~~
```
I didn't read the C++ standard corresponding chapters but the problem is that clang don't like casting const pointer to non const pointer event to `void*`.
So first only `const void*` static cast is called and after the const is removed by const_cast. Probably there is a template magic that can remove const from template parameter but I don't know it.